### ISSUE TITLE
Else clause for automations

### ISSUE THREAD

**Issue by OnFreund** [home-assistant/architecture#378]
## Context

There are a lot of use cases that require one sequence of actions if a condition is true and different one if it's false. For example:
* Checking if a person is authorized to perform an action, and then either performing it or notifying someone of an attempt.
* Performing a different action based on whether it's a weekday or a weekend, home or away, etc...

The existing methods of performing this have their limitations. The easiest way is creating two automations with the same trigger and opposite conditions (this is now easier thanks to https://github.com/home-assistant/core/pull/34624). This means repeating the conditions, which is both not DRY and, more importantly, could also in some cases lead to both automations running (or not running) since the conditions are evaluated twice.
One can avoid repeating the conditions by use of templates and `input_*`, but that could be cumbersome, requires redundant entities, and is still subject to potential timing issues.
Other approaches could include a service template to run two different scripts, but that means the conditions now need to run as actions instead of conditions. I'm sure there are other approaches as well, each with its limitation.

Keeping in mind that we don't want to turn automations into a general purpose programming language, I wanted to propose a relatively simple mechanism, that is able to achieve quite a lot.

## Proposal

I propose adding an else clause to automation, which will be defined in a similar way to the existing actions, but will run when the condition is _false_, rather than when it's _true_.
This covers simple use cases easily, and more advanced cases can be handled through firing events and triggering multiple automations. 

For example:
```yaml
trigger:
  ...
condition:
  - condition: state
    entity_id: binary_sensor.weekday
    state: 'on'
action:
  <actions to perform on a weekday>
else_action:
  <actions to perform during the weekend>
```

## Consequences

It's easier and less error prone to create automations that have if/else logic. Since the else clause is optional, existing automations aren't affected.

**Comment by balloob** [comment]
I think step 1 should be to add else logic to scripts, since scripts and the action part of automations are executed using the same code. Once that is in place, we might think differently about automations :) 

A discussion is currently ongoing here https://github.com/home-assistant/core/pull/35133 but I think that should be moved to arch. 

I am 👍 on adding `else`.

**Comment by OnFreund** [comment]
OK, adding this to scripts allows a lot of use cases, perhaps even more than my proposal. I wanted to start small and intuitive - a single else clause per automation, but if we're ok taking a more bold approach I'm all for it.

**Comment by thomasloven** [comment]
If you don't mind, I think continuing the discussion here would make sense.

The PR linked above adds an `else` option to the `condition` step of script/automation.
I'm not convinced that's the best way since it inverts the flow from what you normally expect.

```yaml
- condition: ...
  else:
    - condition failed
- condition passed
```
rather than
```
if
  then
else
```

The pros of this method is that it simply replaces the default behavior of a failed condition - to stop execution.
Perhaps this problem can be "fixed" by renaming the `else` parameter to `on_fail` or something...

**Comment by pvizeli** [comment]
The perfect solution I think would be to have multiple ` sequence`:

```yaml
script:
  example_script:
    sequence:
      - condition: state
         entity_id: device_tracker.paulus
         state: 'home'
      - service: notify.notify
        data:
          message: 'Turned on the ceiling light!'
    sequence:
      - condition: state
         entity_id: device_tracker.paulus
         state: 'away'
      - service: notify.notify
        data:
          message: 'Turned on the ceiling light!'
```

**Comment by OnFreund** [comment]
Yes, I agree that `on_fail` or something similar would be more representative in that case. Overall, I think these are both legitimate ways of achieving something similar.  I think a single else clause for an automation is a bit more intuitive, especially for users without programming background, but an `on_fail` approach can achieve more use cases without resorting to multiple automations.

**Comment by pvizeli** [comment]
Yeah, I like the idea from @thomasloven, it allow not too complex deeps as my variant, but it work out of the box everywhere and is fine for yaml.

**Comment by OnFreund** [comment]
@pvizeli my main concern with the multiple sequences approach is that the conditions are evaluated multiple times, which can potentially lead to confusing results. In your example, the device tracker can change from `home` to `away` between the sequences (or vice versa) and both services will be called (or neither). Unless the entire state machine is frozen during script execution..

**Comment by OnFreund** [comment]
To rephrase - I think there needs to be an atomic way to evaluate a condition and then branch out accordingly.

**Comment by balloob** [comment]
Maybe a new type `branch` with a list of condition + sequence. The first condition to pass will be executed.

**Comment by balloob** [comment]
The nice thing about a `branch` type is that you also can add generic code to be executed after either branch is done

**Comment by thomasloven** [comment]
I like it. That's a very good point with the code after any branch.

```yaml
- branch:
  - condition: state
    entity_id: input_binary.guest_mode
    state: 'on'
    sequence:
      - service: system_log.write
        data:
          message: There are guest
  - condition: state
    entity_id: input_binary.guest_mode
    state: 'off'
    sequence:
      - service: system_log.write
        data:
          message: Social distancing
  - sequence:
    # No condition -> Always evaluates to true, i.e. an else-clause
      - service: system_log.write
        data:
          message: Something went horribly wrong!
```

**Comment by balloob** [comment]
I would prefer if we would not extend condition schemas but instead have condition and sequence live in a dictionary. This will make validation easier as we don't have to make copies of the dict to pass it to existing validations. 

```yaml
- branch:
  - condition:
      condition: state
      entity_id: input_binary.guest_mode
      state: 'on'
    sequence:
      - service: system_log.write
        data:
          message: There are guest
```

**Comment by OnFreund** [comment]
@balloob would the state machine be frozen between evaluation of the different condition clauses? Otherwise it wouldn't be possible to achieve atomic branching and ensure that one and only one branch is true.

**Comment by elupus** [comment]
@OnFreund As long as there is no async await in the code path it will be frozen.

**Comment by PiotrMachowski** [comment]
I really like suggestion made by @balloob, it can be also extended to incorporate changes suggested by @thomasloven:
```yaml
- branch:
  - condition:
      condition: state
      entity_id: input_select.some_value
      state: state_1
    sequence:
      - service: system_log.write
        data:
          message: State is one
  - condition:
      condition: state
      entity_id: input_select.some_value
      state: state_2
    sequence:
      - service: system_log.write
        data:
          message: State is two
  default_sequence:
    - service: system_log.write
      data:
        message: State has unexpected value
```
It would be something similar to switch statement and pattern matching known from programming languages.

**Comment by OnFreund** [comment]
@elupus cool, that should take care of conditions (I don't believe conditions can have side effects), but we need to ensure that the sequences don't run between the conditions, because those would most likely have awaits in them.
So first run all conditions in one go, then run sequences for conditions that evaluated to true.

@PiotrMachowski I really like the `default_sequence`, I think it opens up a lot of options, and also makes the binary case (one sequence for true, another one for false) really easy.

**Comment by PiotrMachowski** [comment]
> So first run all conditions in one go, then run sequences for conditions that evaluated to true.
I think that it would be better to run only a sequence for a first passed condition (like mentioned above by @balloob)

**Comment by OnFreund** [comment]
Oh, if only the first true condition takes then you're right. (assuming conditions can indeed have no side effects).
In any case, I really like the `default_sequence` suggestion. I vote for [the syntax in your comment](https://github.com/home-assistant/architecture/issues/378#issuecomment-623911800).

**Comment by Jc2k** [comment]
Would this support nesting? Contrived example for syntax of what I mean.

```yaml
- branch:
  - condition:
      condition: state
      entity_id: input_binary.guest_mode
      state: 'on'
    sequence:
    - branch:
      - condition:
          condition: state
          entity_id: binary_sensor.is_house_cold
          state: 'on'
        sequence:
          - service: system_log.write
            data:
              message: There are guests and it is cold
```

**Comment by thomasloven** [comment]
Most likely.

**Comment by pvizeli** [comment]
The `default_sequence` could be pretty hard to implement with how scripts work. I would suggest to start simple with the branch concept and later add the default. Step by Step

**Comment by OnFreund** [comment]
I believe `default_sequence` is the most important part of this (in fact, you could potentially implement everything else with it using nested branches).

**Comment by PiotrMachowski** [comment]
@pvizeli Could it be possible to make `default_sequence` internally just an alias for a condition that is always `true`?

**Comment by balloob** [comment]
if we're going to execute first branch to match, we could just execute first branch that has no condition specified. 

And yes, these can be nested. A sequence doesn't know what it's executing, so it can be yet another branch. It won't be fun to keep track of which branch you're editing though 🤷

**Comment by OnFreund** [comment]
OK, sounds like we're converging to a good outcome.
An example of where we are now:

```yaml
- branch:
  - condition:
      condition: state
      entity_id: input_select.some_value
      state: state_1
    sequence:
      - service: system_log.write
        data:
          message: State is one
  - condition:
      condition: state
      entity_id: input_select.some_value
      state: state_2
    sequence:
      - service: system_log.write
        data:
          message: State is two
      - branch:
        - condition:
            condition: state
            entity_id: binary_sensor.some_value
            state: 'on'
          sequence:
            - service: system_log.write
              data:
                message: Nested conditions work but are not encouraged
  - sequence: # no condition acts as a fallback - it's always true
    - service: system_log.write
      data:
        message: State has unexpected value
```

**Comment by thomasloven** [comment]
This would also pretty much require some way to stop a script execution, so you can stop after some branches, but continue after others.

Simplest way would be to add a 
```
- condition: false
```
but that's not very user friendly.

**Comment by thomasloven** [comment]
My pr has been updated to work with the new syntax: https://github.com/home-assistant/core/pull/35133

**Comment by OnFreund** [comment]
@thomasloven what if conditions break out of the current sequence, but don't stop script execution altogether? I'm not entirely sure about current behavior, but if the following two statements are true, then it's a backwards compatible solution:
1. The only way to nest sequences today is by calling a separate script.
2. A condition only breaks out of the currently executing script, not its caller.

**Comment by thomasloven** [comment]
I'm talking about the case:
```
if A:
  Do someting if A
elif B:
  Do something if B
else:
  Do something else
  exit

Do something common for A and B
```

Without the exit statement, the common action would need to be repeated in the A and B branches.

**Comment by OnFreund** [comment]
I think exit is a nice-to-have, but not strictly required. The "do something common" part can be achieved with repetition, like you mentioned.

**Comment by OnFreund** [comment]
BTW I really like the use of `if` in your syntax, instead of the overloaded `condition` syntax.

**Comment by balloob** [comment]
@thomasloven "do something common for A/B" could be extracted into a script. I don't think that we should try to replicate in linear YAML what Node Red does, it is too complicated and no one will end up using it.

**Comment by OnFreund** [comment]
@balloob are my assumptions in https://github.com/home-assistant/architecture/issues/378#issuecomment-624458394 correct?

**Comment by balloob** [comment]
Correct, a condition stops the execution of a script, not the complete call. So a nested condition will only stop that current branch.

**Comment by thomasloven** [comment]
> @thomasloven "do something common for A/B" could be extracted into a script. I don't think that we should try to replicate in linear YAML what Node Red does, it is too complicated and no one will end up using it.

You're right. My example could be restructured to work even without early return.

**Comment by gjbadros** [comment]
Could we call this "choose" instead of branch?  In general, I find the XSLT standard to be a good go-to inspiration for declarative specification of programming (and it, itself, is quite Lisp/Scheme inspired, but uses XML syntax instead of parens vs. YAML that we use):

https://developer.mozilla.org/en-US/docs/Web/XSLT/Element/choose

**Comment by balloob** [comment]
I agree that `choose` is more understandable for the non-programmer user.

**Comment by gjbadros** [comment]
One other thing that maybe we could clean up if there's not a good reason for it: I've always found the "conditions" tag under conjunctions and disjunctions confusing and unnecessary.  Why don't "condition: and/or" just have a list as a child?  E.g., can we write:

```
condition:
  condition: and
    - condition: state
        ....
    - condition: template
        ....
```

instead of:
```
condition:
  condition: and
    conditions:  # why do we need this?
    - condition: state
        ....
    - condition: template
        ....
```

Maybe there's some nuance I'm missing, but maybe we could deprecate the conditions key and make those work with just a list of the conditions?

**Comment by thomasloven** [comment]
That's impossible with normal data structures.
Basically, you have two ways to store data - a list and a dictionary.
Lists are just lists of stuff
```yaml
- 1
- 2
- 3
```
Dictionaries consist of a number of key-value pairs, where the key is uniqie:
```
name: Thomas
surname: Lovén
country: Sweden
```

A dictionary value can be a list and lists can consist of dictionaries (in your example the `conditions:` key has a list of conditions as its value and each item is a dictionary) but the structures cannot be mixed in the same "level".

**Comment by gjbadros** [comment]
Eek, yea, the example I tested in a parser was actually:

condition:
  condition_and:
    - condition: state
        ....
    - condition: template
        ....

But I suppose the better example is for disjunctions since, IIUC, the list of conditions under the topmost condition are by default treated as a conjunction. 

If we were willing to do _-separated condition_* keys for brevity, condition_not and condition_template would be great candidates, too:

condition:
  condition_template: '{{ ... }}'

condition:
  condition_not:
     - ....
     - ....

Not sure how palatable that approach is to folks, especially since it'd be a change and it's not clear that it's all that much better.

**Comment by OnFreund** [comment]
@gjbadros this is getting off topic. Maybe open a new arch issue?

**Comment by gjbadros** [comment]
dropping it; my other arch issues are more important to me to get feedback on.

**Comment by pnbruckner** [comment]
Coincidentally I started thinking about what to do next with scripts and was toying with some variations of an if construct (and others for looping) and ran across this conversation. I kind of prefer this over the more traditional if, especially since it fits better with YAML. I'd probably use `default`, though:
```yaml
- choose:
  - condition:
    # One or more conditions that are AND'd
    sequence:
    # Sequence that executes if associated condition is first to evaluate to true
  - # Repeat the above as many times as desired
  default: # Optional
  # Sequence that executes if none of the conditions evaluate to true
```
I know it's a bit off topic, but I've started to work on the implementation of the following new features for looping:
```yaml
- while:
  # One or more conditions that are AND'd
  sequence:
  # Sequence that executes repeatedly while above conditions evaluate to true
```
and:
```yaml
- repeat: COUNT # template accepted here as well, rendered once each time script reaches this step
  sequence:
  # Sequence that executes repeatedly the number of times determined by repeat value above
```
Although one _can_ do looping now with a single, `parallel` mode script, it actually causes recursion which is probably less than optimal. I like the idea of native looping constructs. I'll start on PRs for those unless the general consensus says otherwise. And @thomasloven, I'm happy to take over the work on this `choose` option if you like (again, assuming we've reached consensus here.)

**Comment by thomasloven** [comment]
RE: taking over `choose`: Please, feel free to do so. And feel free to build on home-assistant/core#35133 or to throw it all away and do it *right* instead.
The syntax you propose will not work, though, since it mixes lists and dictionaries - see comments marked as off-topic comments.

As mentioned in that PR, I also had some thoughts on looping and on script local variables.
Maybe that should be put in a new arch issue, though. Or this be renamed?

**Comment by pnbruckner** [comment]
> The syntax you propose will not work, though, since it mixes lists and dictionaries

Each of the three I suggested above are perfectly valid. E.g., The "choose" step is a dictionary with two keys: `choose` & `default`, the value of each being a list. And the elements of the `choose` list are themselves dictionaries with two keys: `condition` & `sequence`, and the values for each of those are also lists. For "while", its a dictionary with two keys: `while` & `sequence`, again where the value of each is a list. And for "repeat", it's a dictionary with two keys: `repeat` & `sequence`. The value of `repeat` is a number or string (which can be a template), and the value of `sequence` is a list. In fact, I've already begun the implementation of `repeat` and it's working well with regards to the schema and created objects.

I definitely looked at what you did in 35133 and it was a very good start. In fact, I had originally discarded the idea of "sub-blocks", thinking the implementation didn't lend itself to that. But I obviously had blinders on! 😄 In my "repeat" implementation I'm using that idea -- i.e., creating another `Script` object for each sub-block. So thanks for that!

I'll review your PR again in more detail to better understand what you were thinking re: looping & local variables.

Regarding continuing and/or starting a new architecture topic, I leave that to others. I think this was a very good discussion, but it seems to have reached its natural conclusion. And for repeat & while, those seem straight forward enough that I think they can be handled via PRs. But that's just my opinion.

**Comment by OnFreund** [comment]
@pnbruckner thanks for taking this on, that's awesome!
A couple of notes:
1. I want to make sure we don't lose some of the important finer points that are included in @thomasloven's MVP (e.g. [this resolved conversation](https://github.com/home-assistant/core/pull/35133/commits/08cdae6381caca4eeaa8845b6a3778eb78fb7df2#diff-39de9c1a633ea0cc1069d15dc9545377R323))
2. I really like the use of `if` instead of `condition`, since each `if` clause is a list of `condition` clauses. I suggest we keep it - it's less confusing.
3. I really like `default` but there was [some concern about feasible it is to implement](https://github.com/home-assistant/architecture/issues/378#issuecomment-624012720). The following three comments also discuss this and the alternatives, in case it's not feasible.

**Comment by pnbruckner** [comment]
1. Yes, I'm aware of the concern about atomicity. I will definitely keep that in mind as I attempt to implement these new features. Worst case there may be another way to address it. E.g., with something like this:
```yaml
- choose: SEQUENCE
  sequence:
    abc:
    # Sequence
    def:
    # Sequence
    default:
    # Sequence
```
where `SEQUENCE` would typically be a template. Being a single template it would be evaluated atomically, and would specify which sequence to run. The only downside I see to this is it requires the user to understand templates, which many don't. On the other hand, this is sort of an advanced feature, so is that asking much?

2. `if` or `condition`, I can see an argument for either. Doesn't really matter to me.

3. I don't see the difficulty. Seems pretty straightforward to me.

**Comment by OnFreund** [comment]
1. The implementation by @thomasloven managed to achieve atomicity, so I'd rather stick to that than have a cumbersome template and compare its evaluation to various values.
3. Great, that's definitely my preferred syntax.

**Comment by pnbruckner** [comment]
There certainly is an advantage with the generally agreed upon approach. However, there's also an advantage to a single template. Yes, it will likely be relatively large, but all the "logic" would be in one place, and hence possibly easier to understand and maintain, rather than having that "logic" scattered across many, many lines, with each list of conditions possible fairly far apart from each other. Not necessarily the best for understand-ability/maintenance.

**Comment by OnFreund** [comment]
IDK, it feels like replacing switch statements in a general purpose programming language with a large regular expression and branching on its output :)

**Comment by pnbruckner** [comment]
In fact, it would be very similar to a switch statement -- one expression decides which case to run. Not sure why that would be a bad thing. It's definitely atomic, keeps logic in one place, makes it a bit more consistent with my "repeat" & "while" constructs (i.e., the second part is `sequence`), ...

But, I'm only suggesting this as an alternative if the previously agreed on "choose" construct doesn't work out for some reason.

**Comment by OnFreund** [comment]
Yeah, I get the one expression part, what I meant is that complex logic within a template would be akin to complex logic within a regular expression. Templates are great for simple stuff, but quickly become hard to follow when more logic is added.

Anyway, thanks again for picking this up - really appreciated.

**Comment by pnbruckner** [comment]
Again, I know it's a bit off topic, but I believe I have the implementation of the repeat loop completed. I, of course, still have to add tests and update the docs, but if anyone is interested see https://github.com/home-assistant/core/tree/script-repeat-loop.

**Comment by pnbruckner** [comment]
@OnFreund 

> The implementation by @thomasloven managed to achieve atomicity

That's actually not true. At least it doesn't appear so to me. The code that evaluates the conditions, although it does them all in a single loop, contains `await` statements, so I don't think it can be said to be guaranteed to be atomic.

If atomicity is important, then I think maybe my suggestion for using a single template in a switch style statement may be the way to go. And I don't agree with your assessment that that is less clear. E.g., compare:
```yaml
- choose:
  - if:
    - condition: state
      entity_id: device_tracker.me
      state: office
    - condition: numeric_state
      entity_id: sensor.n_guests
      above: 0
    sequence:
    - service: switch.turn_on
      entity_id: switch.inside_lights
  - if:
    - condition: not
      conditions:
      - condition: state
        entity_id: device_tracker.me
        state: home
    - condition: numeric_state
      entity_id: sensor.luminance
      below: 50
    sequence:
    - service: switch.turn_on
      entity_id: switch.front_porch_light
```
to:
```yaml
- choose: >
    {% if not is_state('device_tracker.me', 'home') and states('sensor.n_guests')|int > 0 %}
      not_home
    {% elif is_state('device_tracker.me', 'office') and states('sensor.luminance')|float < 50 %}
      office
    {% endif %}
  sequence:
    not_home:
    - service: switch.turn_on
      entity_id: switch.inside_lights
    office:
    - service: switch.turn_on
      entity_id: switch.front_porch_light
```
I can tell you which I think is easier to understand and maintain, and the decision making is atomic.

Yes, I understand the the first one could be changed to templates as well, and be more compact. The points are 1) whether the decision logic is in one place as opposed to scattered throughout the statement (which gets worse as there are more choices), and 2) whether atomicity can be achieved in the decision logic.

**Comment by gjbadros** [comment]
Can it be done atomically of all of the "when" condition clauses are of
type 'template'?  It seems like we could document that the only way to have
the conditions evaluated against a stable global state is to limit to
conditions of a small number of types (template-only?) and then it'd be the
same as executing the latter that you give but with, IMO, a much
easier-to-read syntax that more closely mirrors other languages like 'cond'
in scheme/lisp and choose/when in XSLT (and other parts of hass).

There are some places where a user might care about the ability to use
different condition types without caring about atomicity especially given a
fall-through default case that could retrigger or otherwise handle the
obscure case of an expectation of a partitioning of cases not being met.
In particular, a when/default where the default is really just the else
clause would still work fine regardless of the kind of condition being used.

**Comment by OnFreund** [comment]
> That's actually not true. At least it doesn't appear so to me. The code that evaluates the conditions, although it does them all in a single loop, contains await statements, so I don't think it can be said to be guaranteed to be atomic.

If I'm reading this correctly, the `await` happens when you've found a `True` condition, after which there's an immediate break and no more conditions evaluate, so I believe it is atomic.

As for the template approach, don't get me wrong - for me, as an advanced user, it's great, and most of my use cases are indeed about a single expression and multiple potential values to compare against. However, I think the barrier to entry with it might be too high, especially when complex logic is required.

**Comment by thomasloven** [comment]
> Each of the three I suggested above are perfectly valid

You are right, of course. Sorry.
RE: the looping and variables, that's not incorporated in that PR other than being mentioned as a future posibility in the description - so don't look too hard :).

RE: atomicity, there are two problems.
1: A condition being true when evaluated, but not when the branch is executed.
2: State changing during evaluation such that cases where at least one is guaranteed to be true are missed.

The first case probably can't be solved, and is also a "problem" for the normal `condition`.
The second case is what's adressed in the last commit of my PR in a way that minimizes its impact. It's not removed entirely, but I honestly don't think it's likely to ever be a problem. Even less so if there's a default branch.

Of course, an alternative approach could be to freeze a copy of the state machine at the beginning of `_async_branch_step` and use the copy for evaluating the conditions, but the real one for running the branch... I guess that would eliminate the second problem.

**Comment by OnFreund** [comment]
> The second case is what's adressed in the last commit of my PR in a way that minimizes its impact. It's not removed entirely, but I honestly don't think it's likely to ever be a problem. Even less so if there's a default branch.

Can you elaborate on what remains non-atomic in this implementation?
(and I agree that #1 is not a concern, and is also true for normal conditions).

**Comment by thomasloven** [comment]
No, you are right. My mind was stuck in pre-emptive threaded execution for some reason.
Case 2 is solved in home-assistant/core#35133.

**Comment by pnbruckner** [comment]
Are you referring to this bit of code:
```yaml
        conditions = [(
            await self._get_condition_config(b["if"]) if b.get("if") else None,
            b["sequence"]
            ) for b in branches]
```
If so, how is that atomic? There's an `await` in the loop.

> Of course, an alternative approach could be to freeze a copy of the state machine at the beginning of _async_branch_step and use the copy for evaluating the conditions

Now that's an idea!

I personally wouldn't expect the state of the system to be the same between the condition evaluation and the branch execution. That's asking a bit much. I wouldn't worry about that.

> minimizes its impact. It's not removed entirely, but I honestly don't think it's likely to ever be a problem

If there's one thing I've learned in over 40 years of embedded programming, _if_ something can happen, it _will_ happen. And this is exactly the sort of thing to make a system misbehave in mysterious ways that are next to impossible to debug/fix.

I'm still not necessarily saying I think it has to be a single template. I do agree that raises the bar for the average user. The idea about getting a frozen copy of the State Machine while evaluating the conditions is potentially a great way to solve the problem. Just not sure how easy that will be to implement since that would probably also affect the condition helper and who knows what else.

Having said all that, I'm wondering why this is considered a problem for this, but not for any other complex, compound set of conditions. If this is really a problem, wouldn't it be true for, say, an automation's condition section, too?

**Comment by OnFreund** [comment]
@pnbruckner that piece of code doesn't evaluate the conditions, but rather just gathers them. The evaluation [happens here](https://github.com/home-assistant/core/pull/35133/files#diff-39de9c1a633ea0cc1069d15dc9545377R326) and is atomic (in the "between condition evaluations" sense, not "covering both conditions and sequences", which I agree isn't needed).

**Comment by pnbruckner** [comment]
Ah, ok, now I see. Thanks! I wasn't reading the code very closely. My bad! Yep, that makes perfect sense now.

**Comment by pnbruckner** [comment]
So just to make sure we're still all on the same page, the choose statement will look like this:
```yaml
- choose:
  - if:
    - condition: ...
    sequence:
    - ...
  - if:
    - condition: ...
    sequence:
    - ...
  # "Default" case leaves out optional if part:
  - sequence:
    - ...
```
Is that right?

**Comment by OnFreund** [comment]
Perfect!

**Comment by pnbruckner** [comment]
Initial implementation complete: https://github.com/home-assistant/core/tree/script-choose-action. Still need to add test and update docs.
